File: python/aurival/http.py
"""The REST lane. One retry policy lives here (SDK-26/SDK-35/SDK-37) so every
caller — `bot.py`, `Context.reply`, pairing — gets it for free. Nothing from
`aiohttp` escapes: every failure becomes one of `aurival.errors`'s exceptions.
"""

from __future__ import annotations

import asyncio
import json
import logging
import os
import urllib.parse
import uuid
from typing import TYPE_CHECKING

import aiohttp

from . import errors
from .cooldown import CooldownTable

if TYPE_CHECKING:
    from .auth import Auth

DEFAULT_HOST = "https://bots.aurival.com"

_AURIVAL_DEBUG_ENV = "AURIVAL_DEBUG"


class Omitted:
    """The third state a card part on an edit or an ack can be in.

    AMENDMENT-07 §9 gives `embeds`/`buttons` three meanings and `None` is
    already spent on two of them: absent from the request body, and the JSON
    `null` that clears the part. `send_message`'s `if embeds:` shortcut cannot
    express that — it folds "clear" into "absent" — so the body builders below
    test against this sentinel instead, and a real `None` reaches the wire as
    `null`.
    """

    __slots__ = ()

    def __repr__(self) -> str:  # pragma: no cover - debugging aid only
        return "OMITTED"


OMITTED = Omitted()


def _card_body(
    text: str | None,
    embeds: list[dict[str, object]] | Omitted | None,
    buttons: list[dict[str, object]] | Omitted | None,
) -> dict[str, object]:
    """The shared body builder for the two routes that take card parts.

    A key is written when its part is present, and only then: `text=None` is
    absent (AMENDMENT-07 §2 makes `text` unclearable, so `None` has only the
    one meaning there) while `text=""` is a real, legal value on both routes.
    `embeds`/`buttons` write `null` on a real `None` and are left out only on
    `OMITTED`.
    """
    body: dict[str, object] = {}
    if text is not None:
        body["text"] = text
    if not isinstance(embeds, Omitted):
        body["embeds"] = embeds
    if not isinstance(buttons, Omitted):
        body["buttons"] = buttons
    return body


def _default_logger() -> logging.Logger:
    """The "aurival" logger, built once at import time. Left alone, an
    unconfigured Python logger already sends WARNING+ to stderr via the
    logging module's last-resort handler — the only thing missing is
    debug/info visibility, which `AURIVAL_DEBUG=1` turns on here (mirrors
    `sdk/js/src/http.ts`'s `defaultLogger()`, same env var name, so it means
    the same thing in both SDKs). Never touches a `Bot(logger=...)` a caller
    supplied explicitly — this only shapes the shared default instance.
    """
    logger = logging.getLogger("aurival")
    if os.environ.get(_AURIVAL_DEBUG_ENV) == "1" and not logger.handlers:
        logger.setLevel(logging.DEBUG)
        handler = logging.StreamHandler()
        handler.setLevel(logging.DEBUG)
        logger.addHandler(handler)
    return logger


_log = _default_logger()

# Both the rate-limit lane and the api_error/transport lane are bounded here
# (SDK-26): this many tries total, first attempt included.
_MAX_ATTEMPTS = 5
_DEFAULT_RETRY_AFTER = 1.0
# A `retry_after` longer than this is not a transient the request should wait
# out (SDK-40). The generic rate limit is retried INSIDE whatever awaited the
# request, which for a reply is the command handler, and the handler holds the
# event's ack open for as long as it runs. A server once answered 600 s; the
# bot then slept ten minutes per command, every unacked event was redelivered
# every 30 s behind it, and the whole chat looked dead. Past this bound the
# error is raised at once and the handler decides.
_MAX_RETRY_AFTER_S = 15.0


def _decode_object(raw: bytes) -> dict[str, object]:
    """A non-JSON body, or JSON that is not an object, is a `ProtocolError`."""
    try:
        data = json.loads(raw) if raw else None
    except (json.JSONDecodeError, UnicodeDecodeError) as exc:
        raise errors.ProtocolError(f"response body was not valid JSON: {exc}") from None
    if not isinstance(data, dict):
        raise errors.ProtocolError("response body was not a JSON object")
    return data


def _envelope_body(envelope: dict[str, object]) -> dict[str, object]:
    inner = envelope.get("error")
    return inner if isinstance(inner, dict) else envelope


def _retry_after_seconds(header: str | None, envelope: dict[str, object]) -> float:
    if header is not None:
        try:
            return max(0.0, float(header))
        except ValueError:
            pass
    body = _envelope_body(envelope)
    value = body.get("retry_after")
    if isinstance(value, (int, float)):
        return max(0.0, float(value))
    return _DEFAULT_RETRY_AFTER


def _backoff_delay(attempt: int) -> float:
    # attempt is 1-based. Plain exponential, capped — the shape doesn't matter
    # much here since every test injects the sleep; being bounded does.
    return min(0.5 * (2 ** (attempt - 1)), 8.0)


class HttpClient:
    def __init__(
        self,
        session: aiohttp.ClientSession,
        host: str,
        auth: Auth | None = None,
        logger: logging.Logger | None = None,
    ) -> None:
        self._session = session
        self._host = host.rstrip("/")
        self._auth = auth
        self._logger = logger or _log
        # Injectable so a retry/backoff test never actually waits.
        self._sleep = asyncio.sleep
        # AMENDMENT-08 §3: the process-memory table `send()`/`reply()`/
        # `edit()`/`ack()` record card/button cooldowns into, so a later
        # `button.pressed` event (ids only) can resolve back to what was
        # attached. One table per `HttpClient`, which is one per `Bot` run —
        # it survives a socket reconnect, since it isn't connection state.
        self.cooldowns = CooldownTable()

    async def request(
        self,
        method: str,
        path: str,
        *,
        body: dict | None = None,
        headers: dict[str, str] | None = None,
        authenticated: bool = True,
        idempotency_key: str | None = None,
        retry_auth: bool = True,
    ) -> dict:
        """Returns the decoded JSON body. Raises only SDK exceptions."""
        base_headers = dict(headers) if headers else {}
        if idempotency_key is not None:
            # Generated once by the caller and reused verbatim across every
            # retry of this call (SDK-30) — never regenerated per attempt.
            base_headers["Idempotency-Key"] = idempotency_key

        url = self._host + path
        auth_retried = False
        attempt = 0

        while True:
            attempt += 1
            attempt_headers = dict(base_headers)
            if authenticated:
                if self._auth is None:
                    raise errors.AurivalError(
                        "request() needs an authenticated Auth but none was given",
                    )
                token = await self._auth.token()
                attempt_headers["Authorization"] = f"Bearer {token}"

            try:
                async with self._session.request(
                    method, url, json=body, headers=attempt_headers
                ) as resp:
                    status = resp.status
                    raw = await resp.read()
                    retry_after_header = resp.headers.get("Retry-After")
            except (aiohttp.ClientError, asyncio.TimeoutError, OSError) as exc:
                if attempt < _MAX_ATTEMPTS:
                    await self._sleep(_backoff_delay(attempt))
                    continue
                raise errors.TransportError(str(exc)) from None

            if status < 400:
                # A 204 No Content (typing, reactions, delete) has no body to
                # decode — an empty dict, not a ProtocolError (R3). Gated on
                # the status itself, not merely an empty body: an empty-bodied
                # 200 (should one ever occur on an existing endpoint) still
                # raises ProtocolError exactly as it did on 0.1.8.
                if status == 204 and not raw:
                    return {}
                return _decode_object(raw)

            envelope = _decode_object(raw)
            retry_after = _retry_after_seconds(retry_after_header, envelope)
            exc = errors.from_envelope(envelope, status=status, retry_after=retry_after)

            if isinstance(exc, errors.AuthenticationError):
                expired_or_invalid = exc.code in ("access_token_expired", "access_token_invalid")
                if (
                    authenticated
                    and expired_or_invalid
                    and retry_auth
                    and not auth_retried
                    and self._auth is not None
                ):
                    auth_retried = True
                    await self._auth.refresh()
                    continue
                raise exc

            if isinstance(exc, errors.RateLimitError):
                # Only the generic `rate_limited` retries here. `pair_rate_limited`
                # and `sync_rate_limited` are owned by their callers (auth.pair,
                # bot.py's background sync per SDK-35) — retrying them here would
                # make sync_commands swallow SyncRateLimited, which it must not.
                if type(exc) is errors.RateLimited and attempt < _MAX_ATTEMPTS:
                    if retry_after > _MAX_RETRY_AFTER_S:
                        self._logger.warning(
                            "rate limited for %.0fs, longer than the %.0fs this client will wait;"
                            " raising",
                            retry_after,
                            _MAX_RETRY_AFTER_S,
                        )
                        raise exc
                    self._logger.warning("rate limited, retrying after %.1fs", retry_after)
                    await self._sleep(retry_after)
                    continue
                raise exc

            if isinstance(exc, errors.APIError):
                if attempt < _MAX_ATTEMPTS:
                    await self._sleep(_backoff_delay(attempt))
                    continue
                raise exc

            # invalid_request_error, permission_error: never retried.
            raise exc

    async def send_message(
        self,
        chat: str,
        text: str,
        *,
        idempotency_key: str,
        mentions: list[dict[str, str]] | None = None,
        embeds: list[dict[str, object]] | None = None,
        buttons: list[dict[str, object]] | None = None,
    ) -> dict:
        body: dict[str, object] = {"chat": chat, "text": text}
        # CONTRACT-V1 §5.0.1: absent and `[]` mean exactly the same thing, so
        # an empty list is omitted rather than sent — matching the js seam,
        # not merely legal per the contract. Same rule for embeds/buttons.
        if mentions:
            body["mentions"] = mentions
        if embeds:
            body["embeds"] = embeds
        if buttons:
            body["buttons"] = buttons
        return await self.request(
            "POST",
            "/v1/messages",
            body=body,
            idempotency_key=idempotency_key,
        )

    async def set_typing(self, chat: str, is_typing: bool) -> dict:
        return await self.request(
            "POST",
            f"/v1/chats/{chat}/typing",
            body={"is_typing": is_typing},
            idempotency_key=str(uuid.uuid4()),
        )

    async def edit_message(
        self,
        msg: str,
        text: str | None = None,
        *,
        embeds: list[dict[str, object]] | Omitted | None = OMITTED,
        buttons: list[dict[str, object]] | Omitted | None = OMITTED,
    ) -> dict:
        """`PATCH /v1/messages/{msg}` with whichever card parts are present
        (AMENDMENT-07 §2). A part left `OMITTED` keeps whatever the message
        already has; `None` goes on the wire as `null` and clears it. Unlike
        `send_message`, an empty list is never a reason to drop the key — the
        caller has already turned `embeds=[]` into the clearing `None`."""
        body = _card_body(text, embeds, buttons)
        return await self.request(
            "PATCH",
            f"/v1/messages/{msg}",
            body=body,
            idempotency_key=str(uuid.uuid4()),
        )

    async def delete_message(self, msg: str) -> dict:
        return await self.request(
            "DELETE",
            f"/v1/messages/{msg}",
            idempotency_key=str(uuid.uuid4()),
        )

    async def set_reaction(self, msg: str, emoji: str) -> dict:
        return await self.request(
            "PUT",
            f"/v1/messages/{msg}/reactions/{urllib.parse.quote(emoji, safe='')}",
            idempotency_key=str(uuid.uuid4()),
        )

    async def unset_reaction(self, msg: str, emoji: str) -> dict:
        return await self.request(
            "DELETE",
            f"/v1/messages/{msg}/reactions/{urllib.parse.quote(emoji, safe='')}",
            idempotency_key=str(uuid.uuid4()),
        )

    async def ack_interaction(
        self,
        interaction: str,
        text: str | None = None,
        *,
        embeds: list[dict[str, object]] | Omitted | None = OMITTED,
        buttons: list[dict[str, object]] | Omitted | None = OMITTED,
        cooldown_retry_after_ms: int | None = None,
    ) -> dict:
        """`POST /v1/interactions/{interaction}/ack`, optionally carrying the
        replacement card (AMENDMENT-07 §3). With no parts at all the body is
        `None`, not `{}`, so a bare ack is byte-identical on the wire to the
        one 0.5.0 sent: `request` passes `json=body` straight to aiohttp, and
        `json=None` writes no entity body and no content-type.

        `cooldown_retry_after_ms` (AMENDMENT-08 §5.1) is the SDK's own
        automatic cooldown ack, never a developer-facing parameter —
        `ButtonContext.ack()` doesn't expose it. Mutually exclusive with
        `text`/`embeds`/`buttons` on the wire (`cooldown_with_body`); the SDK
        enforces that by construction, only ever calling this with one or the
        other, never both."""
        if cooldown_retry_after_ms is not None:
            body: dict[str, object] = {"cooldown": {"retry_after_ms": cooldown_retry_after_ms}}
        else:
            body = _card_body(text, embeds, buttons)
        return await self.request(
            "POST",
            f"/v1/interactions/{interaction}/ack",
            body=body or None,
            idempotency_key=str(uuid.uuid4()),
        )

    async def list_members(self, chat: str, cursor: str | None = None) -> dict:
        path = f"/v1/chats/{chat}/members"
        if cursor:
            path += f"?cursor={urllib.parse.quote(cursor, safe='')}"
        return await self.request("GET", path)

    async def sync_commands(self, bot: str, commands: list[dict[str, str]]) -> dict:
        return await self.request(
            "PUT",
            f"/v1/bots/{bot}/commands",
            body={"commands": commands},
            idempotency_key=str(uuid.uuid4()),
        )

    async def list_commands(self, bot: str) -> dict:
        return await self.request("GET", f"/v1/bots/{bot}/commands")

    def gateway_url(self) -> str:
        if self._host.startswith("https://"):
            base = "wss://" + self._host[len("https://") :]
        elif self._host.startswith("http://"):
            base = "ws://" + self._host[len("http://") :]
        else:
            base = self._host
        return base + "/v1/gateway"

File: js/src/http.ts
/**
 * The REST lane. One retry policy lives here (SDK-26/SDK-35/SDK-37) so every
 * caller — `bot.ts`, `Context.reply`, pairing — gets it for free. Nothing from
 * `fetch` escapes: every failure becomes one of `./errors.js`'s errors.
 */

import { randomUUID } from 'node:crypto';
import * as errors from './errors.js';
import type { Auth } from './auth.js';

export const DEFAULT_HOST = 'https://bots.aurival.com';

// Both the rate-limit lane and the api_error/transport lane are bounded here
// (SDK-26): this many tries total, first attempt included.
const MAX_ATTEMPTS = 5;
// A `retry_after` longer than this is not a transient the request should wait
// out (SDK-40). The generic rate limit is retried INSIDE whatever awaited the
// request, which for a reply is the command handler, and the handler holds the
// event's ack open for as long as it runs. A server once answered 600 s; the
// bot then slept ten minutes per command, every unacked event was redelivered
// every 30 s behind it, and the whole chat looked dead. Past this bound the
// error is thrown at once and the handler decides.
const MAX_RETRY_AFTER_MS = 15_000;
const DEFAULT_RETRY_AFTER_MS = 1000;
const REQUEST_TIMEOUT_MS = 60_000;

/**
 * Node has no stdlib logger, so this is the seam Python's `logging.Logger`
 * occupies. `defaultLogger` mirrors an unconfigured `logging.getLogger`:
 * warnings and errors reach stderr, debug and info are dropped unless
 * `AURIVAL_DEBUG` is set.
 */
export interface Logger {
  debug(message: string, ...args: unknown[]): void;
  info(message: string, ...args: unknown[]): void;
  warn(message: string, ...args: unknown[]): void;
  error(message: string, ...args: unknown[]): void;
}

function emit(level: string, message: string, args: unknown[]): void {
  process.stderr.write(`aurival ${level}: ${message}${args.length ? ' ' + args.join(' ') : ''}\n`);
}

export function defaultLogger(): Logger {
  const verbose = Boolean(process.env['AURIVAL_DEBUG']);
  return {
    debug: (m, ...a) => {
      if (verbose) emit('debug', m, a);
    },
    info: (m, ...a) => {
      if (verbose) emit('info', m, a);
    },
    warn: (m, ...a) => emit('warn', m, a),
    error: (m, ...a) => emit('error', m, a),
  };
}

export interface RequestOptions {
  body?: Record<string, unknown> | undefined;
  headers?: Record<string, string> | undefined;
  authenticated?: boolean | undefined;
  idempotencyKey?: string | undefined;
  retryAuth?: boolean | undefined;
}

/** A non-JSON body, or JSON that is not an object, is a `ProtocolError`. */
function decodeObject(raw: string): Record<string, unknown> {
  let data: unknown;
  try {
    data = raw ? JSON.parse(raw) : null;
  } catch (exc) {
    const detail = exc instanceof Error ? exc.message : String(exc);
    throw new errors.ProtocolError(`response body was not valid JSON: ${detail}`);
  }
  if (typeof data !== 'object' || data === null || Array.isArray(data)) {
    throw new errors.ProtocolError('response body was not a JSON object');
  }
  return data as Record<string, unknown>;
}

function envelopeBody(envelope: Record<string, unknown>): Record<string, unknown> {
  const inner = envelope['error'];
  return typeof inner === 'object' && inner !== null && !Array.isArray(inner)
    ? (inner as Record<string, unknown>)
    : envelope;
}

function retryAfterMs(header: string | null, envelope: Record<string, unknown>): number {
  if (header !== null) {
    const parsed = Number(header);
    if (Number.isFinite(parsed)) return Math.max(0, parsed * 1000);
  }
  const value = envelopeBody(envelope)['retry_after'];
  if (typeof value === 'number' && Number.isFinite(value)) return Math.max(0, value * 1000);
  return DEFAULT_RETRY_AFTER_MS;
}

// attempt is 1-based. Plain exponential, capped — the shape doesn't matter much
// here since every test injects the sleep; being bounded does.
function backoffDelayMs(attempt: number): number {
  return Math.min(500 * 2 ** (attempt - 1), 8000);
}

export class HttpClient {
  readonly #host: string;
  readonly #auth: Auth | null;
  readonly #logger: Logger;
  /** Injectable so a retry/backoff test never actually waits. */
  sleep: (ms: number) => Promise<void> = (ms) =>
    new Promise((resolve) => {
      setTimeout(resolve, ms);
    });

  constructor(host: string, auth: Auth | null = null, logger?: Logger) {
    this.#host = host.replace(/\/+$/, '');
    this.#auth = auth;
    this.#logger = logger ?? defaultLogger();
  }

  /** Returns the decoded JSON body. Throws only SDK errors. */
  async request(
    method: string,
    path: string,
    options: RequestOptions = {},
  ): Promise<Record<string, unknown>> {
    const baseHeaders: Record<string, string> = { ...(options.headers ?? {}) };
    if (options.idempotencyKey !== undefined) {
      // Generated once by the caller and reused verbatim across every retry of
      // this call (SDK-30) — never regenerated per attempt.
      baseHeaders['Idempotency-Key'] = options.idempotencyKey;
    }
    const authenticated = options.authenticated ?? true;
    const retryAuth = options.retryAuth ?? true;

    const url = this.#host + path;
    let authRetried = false;
    let attempt = 0;

    for (;;) {
      attempt += 1;
      const headers: Record<string, string> = { ...baseHeaders };
      if (authenticated) {
        if (this.#auth === null) {
          throw new errors.AurivalError('request() needs an authenticated Auth but none was given');
        }
        headers['Authorization'] = `Bearer ${await this.#auth.token()}`;
      }

      let status: number;
      let raw: string;
      let retryAfterHeader: string | null;
      try {
        const init: RequestInit = {
          method,
          headers,
          signal: AbortSignal.timeout(REQUEST_TIMEOUT_MS),
        };
        if (options.body !== undefined) {
          headers['content-type'] = 'application/json';
          init.body = JSON.stringify(options.body);
        }
        const resp = await fetch(url, init);
        status = resp.status;
        raw = await resp.text();
        retryAfterHeader = resp.headers.get('Retry-After');
      } catch (exc) {
        // NOTHING about `exc` is logged or attached (SDK-33): a fetch failure
        // carries a `cause` chain that can reach back to the request, and this
        // request has an Authorization header on it. Only the bare message.
        const detail = exc instanceof Error ? exc.message : String(exc);
        if (attempt < MAX_ATTEMPTS) {
          await this.sleep(backoffDelayMs(attempt));
          continue;
        }
        throw new errors.TransportError(detail);
      }

      // 204 + an empty body decodes to `{}` — several new mutating routes
      // (typing, reactions, delete) answer this way. A 204 that (against
      // spec) carries a JSON body decodes it rather than silently dropping
      // it, gating on the body being empty rather than on status alone —
      // mirrors http.py's identical rule (merge-gate senior review). In
      // practice `raw` is ALWAYS '' here: the WHATWG fetch spec forces a
      // null body for every 204/205/304 response, so `resp.text()` above
      // already discarded whatever bytes the server actually sent before
      // this line runs (verified against Node's own `fetch`). This branch
      // stays correct for whatever `raw` it is given regardless — python's
      // http.py reads raw socket bytes and is not subject to this
      // fetch-specific stripping, so the decode path is live there. An empty
      // 200 (or any other non-204 success) is still a ProtocolError: every
      // other status decodes unconditionally, so an empty body only becomes
      // {} through the 204 branch, exactly as before this change.
      if (status === 204) return raw === '' ? {} : decodeObject(raw);
      if (status < 400) return decodeObject(raw);

      const envelope = decodeObject(raw);
      const retryAfter = retryAfterMs(retryAfterHeader, envelope);
      const exc = errors.fromEnvelope(envelope, { status, retryAfter: retryAfter / 1000 });

      if (exc instanceof errors.AuthenticationError) {
        const expiredOrInvalid =
          exc.code === 'access_token_expired' || exc.code === 'access_token_invalid';
        if (authenticated && expiredOrInvalid && retryAuth && !authRetried && this.#auth !== null) {
          authRetried = true;
          await this.#auth.refresh();
          continue;
        }
        throw exc;
      }

      if (exc instanceof errors.RateLimitError) {
        // Only the generic `rate_limited` retries here. `pair_rate_limited` and
        // `sync_rate_limited` are owned by their callers (auth.pair, bot.ts's
        // background sync per SDK-35) — retrying them here would make
        // syncCommands swallow SyncRateLimited, which it must not.
        if (exc.constructor === errors.RateLimited && attempt < MAX_ATTEMPTS) {
          if (retryAfter > MAX_RETRY_AFTER_MS) {
            this.#logger.warn(
              `rate limited for ${(retryAfter / 1000).toFixed(0)}s, longer than the ${(MAX_RETRY_AFTER_MS / 1000).toFixed(0)}s this client will wait; throwing`,
            );
            throw exc;
          }
          this.#logger.warn(`rate limited, retrying after ${(retryAfter / 1000).toFixed(1)}s`);
          await this.sleep(retryAfter);
          continue;
        }
        throw exc;
      }

      if (exc instanceof errors.APIError) {
        if (attempt < MAX_ATTEMPTS) {
          await this.sleep(backoffDelayMs(attempt));
          continue;
        }
        throw exc;
      }

      // invalid_request_error, permission_error: never retried.
      throw exc;
    }
  }

  /**
   * `replyTo` is a `msg_` reference and is OMITTED when absent — the field is
   * optional on the wire (CONTRACT-V1 §5.0) and a `null` there is a
   * `parameter_invalid` waiting to happen.
   */
  async sendMessage(
    chat: string,
    text: string,
    idempotencyKey: string,
    replyTo?: string | null,
    mentions?: Array<{ user: string }>,
  ): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = { chat, text };
    if (replyTo != null && replyTo !== '') body['reply_to'] = replyTo;
    if (mentions !== undefined && mentions.length > 0) body['mentions'] = mentions;
    return this.request('POST', '/v1/messages', { body, idempotencyKey });
  }

  /** `POST .../typing` body is `{"is_typing": true|false}` — never `state`. */
  async setTyping(chat: string, isTyping: boolean): Promise<void> {
    await this.request('POST', `/v1/chats/${chat}/typing`, {
      body: { is_typing: isTyping },
      idempotencyKey: randomUUID(),
    });
  }

  async editMessage(message: string, text: string): Promise<Record<string, unknown>> {
    return this.request('PATCH', `/v1/messages/${message}`, {
      body: { text },
      idempotencyKey: randomUUID(),
    });
  }

  async deleteMessage(message: string): Promise<void> {
    await this.request('DELETE', `/v1/messages/${message}`, { idempotencyKey: randomUUID() });
  }

  /** The emoji is percent-encoded — it travels in the path, not the body. */
  async setReaction(message: string, emoji: string): Promise<void> {
    await this.request('PUT', `/v1/messages/${message}/reactions/${encodeURIComponent(emoji)}`, {
      idempotencyKey: randomUUID(),
    });
  }

  async unsetReaction(message: string, emoji: string): Promise<void> {
    await this.request('DELETE', `/v1/messages/${message}/reactions/${encodeURIComponent(emoji)}`, {
      idempotencyKey: randomUUID(),
    });
  }

  /** GET, so no idempotency key — reads are never retried-as-a-write. */
  async listMembers(chat: string, cursor?: string): Promise<Record<string, unknown>> {
    const query =
      cursor !== undefined && cursor !== '' ? `?cursor=${encodeURIComponent(cursor)}` : '';
    return this.request('GET', `/v1/chats/${chat}/members${query}`);
  }

  async syncCommands(
    bot: string,
    commands: Array<{ name: string; description: string }>,
  ): Promise<Record<string, unknown>> {
    return this.request('PUT', `/v1/bots/${bot}/commands`, {
      body: { commands },
      idempotencyKey: randomUUID(),
    });
  }

  async listCommands(bot: string): Promise<Record<string, unknown>> {
    return this.request('GET', `/v1/bots/${bot}/commands`);
  }

  gatewayUrl(): string {
    let base: string;
    if (this.#host.startsWith('https://')) base = 'wss://' + this.#host.slice('https://'.length);
    else if (this.#host.startsWith('http://')) base = 'ws://' + this.#host.slice('http://'.length);
    else base = this.#host;
    return base + '/v1/gateway';
  }
}
